feat(mTLS): adds mTLS support to dataplane api-server

EandrewJones opened the pull request:
Closes #50.

### Questions
I am not sure how best to add tests for this. Unit testing the config trivially tests the behavior of Opt which we know works. Unit testing `setup_tls` forces me to mock `read_to_string` in the absence of an actual file to read in and all the ways I've read about doing that in rust feel really kludgie / dirty up the code with dependency injections, etc.

1. Is this all we need to add mTLS support for the api-server?
2. Is there some sort of API contract we need to maintain for this such that I can write a conformance or integration test?
3. What documentation would we like to see for this?

aryan9600 commented on dataplane/api-server/src/config.rs:
```diff
@@ -0,0 +1,61 @@
+    pub enable_tls: bool,
+    #[clap(long, default_value = "false")]
+    pub enable_mtls: bool,
+    pub certificate_authority_root_path: Option<PathBuf>,
```
i might be lacking the required context but is there a reason this is not `server_certificate_authority_root_path`?

EandrewJones replied:
Updating name.

aryan9600 commented on dataplane/api-server/src/config.rs:
```diff
@@ -0,0 +1,61 @@
+
+impl GrpcConfig {
+    pub fn validate(&self) -> Result<(), String> {
+        fn validate_paths(paths: &[(&Option<PathBuf>, &str)]) -> Result<(), String> {
```
is this solely done to avoid a move? asking because i'm not a big fan of inline functions unless they're essential. this can be just its own function that accepts a vector. please correct me if i'm wrong.

EandrewJones replied:
No moving at play here. I generally agree nested functions suck and avoid them. In this case, I felt it wasn't that big of a deal because it's not intended to be used anywhere outside of the validate method. I tend to feel extremely tight coupling is the only excuse for nested functions.

Either way, I moved it out.

aryan9600 commented on dataplane/api-server/src/lib.rs:
```diff
@@ -5,33 +5,78 @@ SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+                tls_config
+                    .server_private_key_path
+                    .as_ref()
+                    .expect("Missing private key path"),
```
lets bubble this error up instead? `async fn start()` already returns a `Result`, so we can just return it up there.

EandrewJones replied:
K.

aryan9600 commented on dataplane/api-server/src/config.rs:
```diff
@@ -0,0 +1,61 @@
+    pub server_certificate_path: Option<PathBuf>,
+    pub server_private_key_path: Option<PathBuf>,
+    pub client_certificate_authority_root_path: Option<PathBuf>,
+    pub client_certificate_path: Option<PathBuf>,
+    pub client_private_key_path: Option<PathBuf>,
```
the below two properties don't seem to be actually used anywhere?

EandrewJones replied:
correct. These are only relevant client side or if this server also acted as a client w.r.t to a different server.

Removing.

aryan9600 commented on dataplane/api-server/src/config.rs:
```diff
@@ -0,0 +1,61 @@
+}
+
+impl GrpcConfig {
+    pub fn validate(&self) -> Result<(), String> {
```
i couldn't find where this is being called. please ignore if clap calls this automatically for us.

EandrewJones replied:
I swore in my reading, I saw clap would use this method for us. But, alas, I cannot find it in docs. I added an explicit call for this now.

If you know of a better way to get clap to perform validation at parse time, please let me known.

aryan9600 commented on dataplane/loader/src/main.rs:
```diff
@@ -20,6 +21,8 @@ use log::{info, warn};
     #[clap(short, long, default_value = "lo")]
     iface: String,
+    #[clap(flatten)]
+    tls_config: GrpcConfig,
```
maybe?

```suggestion
    grpc_tls_config: GrpcConfig,
```

EandrewJones replied:
Good call out.

I have changed the name of the struct to match the function argument: `GrpcConfig` --> `TlsConfig`. There's nothing about the config that makes it unique to Grpc so I don't know why I named it that way originally.

aryan9600 commented on dataplane/api-server/src/config.rs:
```diff
@@ -0,0 +1,61 @@
+use std::path::PathBuf;
+
+#[derive(Debug, Parser, Clone)]
+pub struct GrpcConfig {
```
maybe?

```suggestion
pub struct GrpcTLSConfig {
```

EandrewJones replied:
Changed to TLSConfig

aryan9600 commented:
> Unit testing setup_tls forces me to mock read_to_string in the absence of an actual file to read in

how about generating certs during the test, writing them to temp files and then passing those to `setup_tls`?

> 1. Is this all we need to add mTLS support for the api-server?

this looks good. we can test it using https://github.com/islishude/grpc-mtls-example/blob/main/cmd/client/main.go.

> 2.  Is there some sort of API contract we need to maintain for this such that I can write a conformance or integration test?

we don't have an integration test for the current grpc server. i guess an integration test for the api server would involve spinning up the server and calling each method from an independent client? for testing mTLS specifically, i think we can just get away with only testing one method, since the thing we want to test is actually the TLS connection. something, like the answer to your first question, but coded in a script.

>  3. What documentation would we like to see for this?

we don't really have any docs apart from code comments. i think code comments for now is fine. documentation is an entirely separate conversation that we need to have soon.

EandrewJones commented:
> how about generating certs during the test, writing them to temp files and then passing those to `setup_tls`?

Added.

> > 1. Is this all we need to add mTLS support for the api-server?
> 
> this looks good. we can test it using https://github.com/islishude/grpc-mtls-example/blob/main/cmd/client/main.go.

If we're not treating this as an integration test, where should we add this so it runs in our pipeline? We need the dataplane server running so we can mimic the client, so it seems like it should go into the integration test pipeline (`tests/integration`). However, the makefile states these tests are deprecated, so should I really be adding new tests to the suite?

> > 3. What documentation would we like to see for this?
> 
> we don't really have any docs apart from code comments. i think code comments for now is fine. documentation is an entirely separate conversation that we need to have soon.

Okay.

shaneutt commented on dataplane/api-server/src/config.rs:
```diff
@@ -0,0 +1,51 @@
+    pub enable_tls: bool,
+    #[clap(long, default_value = "false")]
+    pub enable_mtls: bool,
+    pub server_certificate_path: Option<PathBuf>,
+    pub server_private_key_path: Option<PathBuf>,
+    pub client_certificate_authority_root_path: Option<PathBuf>,
```
So in general I've seen a lot of `Option<T>` fields in struct attributes over the years and that's fine in general. What's interesting in this case is that you're not trying to express that these fields are ultimately optional, but instead that they are optional situationally based on the values of other fields. In effect, this means that you have a couple of different _variants_ of a `TLSConfig` that are acceptable.

Given that notion, and the fact that I immediate thought "variants", I think you should consider whether an [Enumeration](https://doc.rust-lang.org/rust-by-example/custom_types/enum.html) might actually be a better fit here, where we would have a `TLSConfig::ServerOnlyTLS` variant and a `TLSConfig::MutualTLS`.

Now... the only problem with this is that I think the standards really don't accept `ServerOnlyTLS` as official nomenclature everywhere they often just say `TLS`, so... that might require some finesse but either way just think this over and let me know what you think.

EandrewJones replied:
Why not just `TLSConfig::TLS` and `TLSConfig::MutualTLS`? Those correspond to the `TLS` and `mTLS` acronyms, respectively, which are standard afaik.

shaneutt replied:
I'm good with those names :+1:

shaneutt commented on dataplane/api-server/src/config.rs:
```diff
@@ -0,0 +1,51 @@
+}
+
+impl TLSConfig {
+    pub fn validate(&self) -> Result<()> {
+        let tls_paths = &[
+            (&self.server_certificate_path, "server_certificate_path"),
+            (&self.server_private_key_path, "server_private_key_path"),
+        ];
+
+        let mtls_paths = &[(
+            &self.client_certificate_authority_root_path,
+            "client_certificate_authority_root_path",
+        )];
+
+        if self.enable_mtls {
+            validate_paths(tls_paths)?;
+            validate_paths(mtls_paths)?;
+        } else if self.enable_tls {
+            validate_paths(tls_paths)?;
+        }
+
+        Ok(())
+    }
+}
+
+fn validate_paths(paths: &[(&Option<PathBuf>, &str)]) -> Result<()> {
+    for (path, name) in paths {
+        if path.is_none() {
+            return Err(anyhow!("Missing required path: {}", name));
+        }
+    }
+    Ok(())
+}
```
If we go with the `Enum` approach outlined above, then none of this Validation would be needed, but we could create [associated functions](https://doc.rust-lang.org/stable/book/ch05-03-method-syntax.html#associated-functions) to [construct](https://rust-unofficial.github.io/patterns/idioms/ctor.html) these variants.

EandrewJones replied:
Is this even necessary? Shouldn't clap handle this for us during parsing? The only reason I was doing this was to enforce the conditional dependencies between fields. The enum renders that irrelevant.

Can't we just pass `opt.tls_config` down to the `setup_tls` function and match on the variant?

shaneutt replied:
Not necessary no, just a thought depending on how you'd like this to end up looking. But the spirit behind it was in-line with the second thing you were suggesting: once we use an enum "validation" does effectively just get outmoded by matching on the variants and applying the appropriate functionality thereafter. :+1:

EandrewJones commented:
DO NOT MERGE. Still needs a test.

EandrewJones commented:
I pushed up a go test that creates certs, spins up the dataplane, and dials the grpc w/ mTLS from a client. The test passes, but I suspect it's a false positive because the dataplane exits almost immediately on my machine due to an error:

```
Error: map error: failed to create map `AYA_LOGS` with code -1

Caused by:
    0: failed to create map `AYA_LOGS` with code -1
    1: Operation not permitted (os error 1)
Error: map error: failed to create map `GATEWAY_INDEXES` with code -1

Caused by:
    0: failed to create map `GATEWAY_INDEXES` with code -1
    1: Operation not permitted (os error 1)
```
I only care about having a valid test at this point and adding it to our CI.

EandrewJones commented on test/integration/dataplane_mtls_test.go:
```diff
@@ -0,0 +1,182 @@
+	return nil
+}
+
+// runDockerImage starts the Docker container with the specified name.
+func runDockerImage(containerName, imageName string) error {
+	cmd := exec.Command(
+		"docker",
+		"run",
+		"--name", containerName,
+		"-d",
+		"-p", "9874:9874",
+		"-v", os.Getenv("PWD")+"/certs:/app/certs",
+		imageName,
+		"mutual-tls",
+		"--server-certificate-path", "/app/certs/server.pem",
+		"--server-private-key-path", "/app/certs/server-key.pem",
+		"--client-certificate-authority-root-path", "/app/certs/root.pem",
+	)
+	return cmd.Run()
+}
```
Not using KTF to spin up the dataplane. I believe that's a problem.

shaneutt replied:
Right, we use KTF specifically for MetalLB which we use for IPAM for our (bpf hijacked) services right now. So for a complete testing environment we're still a bit bound to KTF (or you can manually reproduce the same environment with your own metallb configmap).

EandrewJones commented:
Does the golang controlplane archival have any impact on this? Or do I just
need to stand up KTF before I run these tests.

Best

Evan Jones
Website: www.ea-jones.com


On Wed, Oct 23, 2024 at 4:50 PM Shane Utt ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In test/integration/dataplane_mtls_test.go
> <https://github.com/kubernetes-sigs/blixt/pull/280#discussion_r1813983933>
> :
>
> > +// runDockerImage starts the Docker container with the specified name.
> +func runDockerImage(containerName, imageName string) error {
> +	cmd := exec.Command(
> +		"docker",
> +		"run",
> +		"--name", containerName,
> +		"-d",
> +		"-p", "9874:9874",
> +		"-v", os.Getenv("PWD")+"/certs:/app/certs",
> +		imageName,
> +		"mutual-tls",
> +		"--server-certificate-path", "/app/certs/server.pem",
> +		"--server-private-key-path", "/app/certs/server-key.pem",
> +		"--client-certificate-authority-root-path", "/app/certs/root.pem",
> +	)
> +	return cmd.Run()
> +}
>
> Right, we use KTF specifically for MetalLB which we use for IPAM for our
> (bpf hijacked) services right now. So for a complete testing environment
> we're still a bit bound to KTF (or you can manually reproduce the same
> environment with your own metallb configmap).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/blixt/pull/280#discussion_r1813983933>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJ2T6AMKOYZPALI3TE4QDBTZ5AY2ZAVCNFSM6AAAAABN3I65QGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGOJQHA3TSMJUGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

shaneutt commented:
> Does the golang controlplane archival have any impact on this? Or do I just need to stand up KTF before I run these tests?

The archival shouldn't have an impact: there's a `make build.cluster` target that sends the right flags (and will automatically install `ktf` for you). The Golang integration tests actually were left in `main` despite removing all the other Go code (since the tests in still have value and help us validate our Rust changes until we replace them with Rust tests).

EandrewJones commented:
@shaneutt @aryan9600 This should be ready for final review. I finally have passing tests. Upon `lgtm`, I'll squash my commits.

EandrewJones commented:
Rebased.

Best

Evan Jones
Website: www.ea-jones.com


On Mon, Dec 2, 2024 at 11:26 AM Shane Utt ***@***.***> wrote:

> Woops I still owe you a review! If you wouldn't mind doing a rebase to
> main I'll have some time this week to start a review (I was out the last
> couple weeks sorry for the lag ^_^)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/blixt/pull/280#issuecomment-2512558534>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJ2T6AK7S2D5FHCE5F4WF432DSX7TAVCNFSM6AAAAABN3I65QGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMJSGU2TQNJTGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

EandrewJones commented:
Looks like changes to the docker-build on main break this branches docker-build now. Looks like it's due to the branch name breaking docker syntax (can't include '#' is my guess).

I won't have time to touch this this week. :(

shaneutt commented:
Ok! No worries, no rush.

EandrewJones commented:
You still can review. I don't recall what I've done since you last
reviewed. Maybe the enum variants and the test?

On Tue, Dec 3, 2024, 08:31 Shane Utt ***@***.***> wrote:

> Ok! No worries, no rush.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/blixt/pull/280#issuecomment-2515038556>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJ2T6AN4WBTFQXZM47FBMVD2DXMG3AVCNFSM6AAAAABN3I65QGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMJVGAZTQNJVGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

shaneutt commented on dataplane/api-server/src/lib.rs:
```diff
@@ -5,33 +5,128 @@ SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+    //
+    // Solution: separate gRPC services
+    //
+    // Public server without mTLS for healthchecks
```
If it's OK with you I would like to put a finer point on this comment to make it clear that no dataplane functionality is enabled through this unencrypted endpoint:

```suggestion
    // Public server without TLS (healthchecks ONLY)
```

shaneutt commented on dataplane/api-server/src/lib.rs:
```diff
@@ -5,33 +5,128 @@ SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
 }
+
+pub fn setup_tls(mut builder: Server, tls_config: &Option<TLSConfig>) -> Result<Server> {
+    // TLS implementation drawn from Tonic examples.
+    // See: https://github.com/hyperium/tonic/blob/master/examples/src/tls_client_auth/server.rs
```
I always appreciate when we link to source materials and relevant external documentation, ty :+1:

shaneutt commented on dataplane/api-server/tests/test_setup_tls.rs:
```diff
@@ -0,0 +1,87 @@
+        "setup_tls should succeed with valid self-signed certs"
+    );
+    Ok(())
+}
```
For completeness it would be ideal to cover the negative inverse cases of both these positive test cases. That is to say: these tests currently build up to an expected success result and verify it, we should _also_ cover the failure cases and verify that we get the expected error result.

shaneutt commented on dataplane/loader/src/main.rs:
````diff
@@ -16,12 +17,51 @@ use clap::Parser;
+///
+/// # Running with mutual TLS config:
+/// $ dataplane --iface eth0 mutual-tls --server-certificate-path /path/to/cert --server-private-key-path /path/to/key --client-certificate-authority-root-path /path/to/ca
+/// ```
````
Your taking the time to add documentation here is greatly appreciated :bow: :+1:

shaneutt commented on dataplane/loader/src/main.rs:
```diff
@@ -62,6 +102,7 @@ async fn main() -> Result<(), anyhow::Error> {
         .context("failed to attach the egress TC program")?;
 
     info!("starting api server");
+    info!("Using tls config: {:?}", &opt.tls_config);
```
:+1:

EandrewJones replied:
I don't love the way rust destructures (or doesn't, rather) the config when printing it, but that's a small nit I can live with for now.

shaneutt commented on test/integration/dataplane_mtls_test.go:
```diff
@@ -0,0 +1,118 @@
+//go:build dataplane_tests
```
I'm not going to say no to adding Golang tests here because this PR review has gone on long enough already and I appreciate your efforts and would prefer to get this in and focus on iterating than to try and stop and do a major rework again now. That said, I felt it would be remiss of me not to point out that we'll likely need to rewrite this into Rust tests at some point so just keep that in mind! Feel free to consider this comment resolved at your discretion.

EandrewJones replied:
Yeah, I'm aware. I wrote these in go out of laziness; I drew on the precedent of integration tests written in golang.

shaneutt commented on test/integration/dataplane_mtls_test.go:
```diff
@@ -0,0 +1,118 @@
+	ServerKeyFile  = certFile("server-key.pem")
+)
+
+// TestGRPCClient tests the gRPC client against a running Docker container.
+func TestGRPCClient(t *testing.T) {
```
Is it worth pointing out here (for the future code reader) that this test is verifying our mTLS configuration works?

```suggestion
// TestGRPCClientWithMTLS tests the client connectivity to the dataplane's gRPC API server. Mutual TLS (mTLS)
// is required to connect to the dataplane API server, and so as a part of this test we create a client certificate
// using the same CA the server used in order to establish mutual trust.
func TestGRPCClientWithMTLS(t *testing.T) {
```

Maybe a bit too wordy, so feel free to ignore this specific suggestion, but I think a breadcrumb here that let's future readers better understand that mTLS configuration is effectively being tested here would be good.

shaneutt commented on test/integration/dataplane_mtls_test.go:
```diff
@@ -0,0 +1,118 @@
+	panic("Env var TEST_CERTS_PATH not found. Please specify path to mTLS test certs")
+}
+
+func setupTLSConfig(cfg TLSConfig) (*tls.Config, error) {
```
There's a bunch of optionality present in here, in that it will try and generate certificates if not present, e.t.c.. However (if I'm not misunderstanding) we actually currently expect for all runs of this test that all values are present? If so, and because this is a private function only used here for this purpose, we may actually just want to _fail_ when things are not present instead as this will help us more quickly realize when we've broken our build system. Wdyt?

EandrewJones replied:
This function does not generate certs if they are not absent, it just reads in the certs and converts them into the tls.config struct that grpc expects.

However, it doesn't abort early if paths to CertFile, KeyFile or CAFile are not present. I can add those.

shaneutt reviewed (approved):
Thank you so much for all your hard work making this happen.

I approve in that I think we're just about ready to go and I don't see any major blocking issues, however I do have some comments I would like to see resolved before we merge please. Also we gotta figure out the CI issue, but in any case, I think we're really close. Thank you!

EandrewJones commented on Makefile:
```diff
@@ -216,6 +230,36 @@ lint: ## Lint Rust code
+	TEST_CERTS_PATH=./$(TEST_CERTS_PATH) \
+	BLIXT_DATAPLANE_IMAGE=$(BLIXT_DATAPLANE_IMAGE):$(TAG) \
+	GOFLAGS="-tags=dataplane_tests" go test -race -v ./test/integration/...
+	$(MAKE) test.rmcert
```
I am happy mpt to remove them since this test will mostly run in ephemeral CI environments. But I think we _should keep_ the targets for creating them. We want to have an automated way to create self-signed certs.

shaneutt replied:
Right, exactly what I was getting at. Keep the _targets_ but don't do cleanup automatically here. :+1:

EandrewJones commented:
@shaneutt I addressed all your comments and fixed the docker-build naming issue by switching the workflow to use the PR number instead of branch name since it's guaranteed to comply with docker's naming rules.

The last thing I need to do is squash. Is there a way for me to squash and re-push without triggering the docker-build action? It's extremely expensive/long-running.

shaneutt reviewed (approved):
/approve
/unhold

/cc @aryan9600

shaneutt commented:
Oh wow yeah 70m build time... holy cow.

shaneutt reviewed (approved):
We could always spend more time here, but given the project is still nascent and we do not have end-users for breakages to impact, let's merge this and start using it!

Thanks for all your hard work @EandrewJones :bow: 

/lgtm
